Replacing CudaAsyncBuffer with TArray to improve perf

weixingzhang opened the pull request:
**Description**: Describe your changes.
Using TArray instead of CudaAsyncBuffer to passing tensor meta data(mostly, tensor strides/pitches) to CUDA kernels to improve performance

**Motivation and Context**
 Basically, with CudaAsyncBuffer, the sequence at host side is like below. 
1. fill meta data to CPU Pinned memory
2. call cudaMemcpyAsync to upload these meta data from CPU Pinned memory to GPU memory.
3. launch CUDA kernel.

On device(GPU) side, the copy in #2 won't be executed until the moment when kernel in #3 is about to run on GPU. Usually, host side runs far ahead of device side. So, it is better to do data uploading early(during kernel launch time) on host side rather than until kernel execution to save time on GPU. To achieve it, we introduced TArray by which the data is passed to CUDA kernels by pass-by-value. On BERT-L training, we saw about 3-5% perf improvement.

KeDengMS commented:
" On BERT-L training, we saw about 3-5% perf improvement." I assume it's for non-fused model right? The change does not seem to touch contrib ops

KeDengMS commented on onnxruntime/core/providers/cuda/object_detection/non_max_suppression.cc:
```diff
@@ -94,40 +94,35 @@ Status NonMaxSuppression::ComputeInternal(OpKernelContext* ctx) const {
     ORT_ENFORCE(output != nullptr);
     int64_t* dst = output->MutableData<int64_t>();
-    size_t count = all_selected_indices.size();
+    int32_t count = static_cast<int32_t>(all_selected_indices.size());
```
>all_selected_indices [](start = 41, length = 20)

not sure if the default array size of 8 would be enough. This value depends on selected bounding box, and may pass test but crash in real scene.

KeDengMS commented on onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc:
```diff
@@ -348,14 +348,13 @@ Status CudnnRnnBase<T>::ComputeInternal(OpKernelContext* ctx) const {
-    CudaAsyncBuffer<int32_t> sequence_lens_buffer(this, batch_size);
-    memcpy(sequence_lens_buffer.CpuPtr(), sequence_lens_data, batch_size * sizeof(int32_t));
-    ORT_RETURN_IF_ERROR(sequence_lens_buffer.CopyToGpu());
+    TArray<int32_t> sequence_lens_buffer(batch_size);
```
>batch_size [](start = 41, length = 10)

not safe for batch_size > 8...I doubt if we have enough test case to catch this though.

KeDengMS commented:
In general, I have concerns of regression in this change not caught by existing tests. I'd suggest locally change the default TArray size to 4, which should be enough to cover most tensor shape related ops (most ONNX ops assume <= 4D tensors). If any op tests failed, take a close look to decide whether its CudaAsyncBuffer can be replaced by fixed size TArray; Then we can change default TArray size to 8 again.

KeDengMS commented on onnxruntime/core/providers/cuda/tensor/concat.cc:
```diff
@@ -59,24 +58,20 @@ Status Concat::ComputeInternal(OpKernelContext* ctx) const {
-  axis_dimension_input_output_mapping_gpu.CopyToGpu();
-  concat_sizes_range_gpu.CopyToGpu();
-  input_ptr.CopyToGpu();
+  TArray<int64_t> concat_sizes_gpu(concat_sizes);
```
>concat_sizes [](start = 35, length = 12)

concat takes variadic inputs, so the default TArray size may not be enough for some models.

KeDengMS commented on onnxruntime/core/providers/cuda/tensor/split.cc:
```diff
@@ -47,8 +47,7 @@ Status Split::ComputeInternal(OpKernelContext* ctx) const {
 
-  CudaAsyncBuffer<void*> output_ptr(this, num_outputs);
-  gsl::span<void*> output_ptr_span = output_ptr.CpuSpan();
+  TArray<void*> output_ptr(num_outputs);
```
>num_outputs [](start = 27, length = 11)

Split generates variadic outputs, so fixed size TArray may not be enough.

weixingzhang commented:
> " On BERT-L training, we saw about 3-5% perf improvement." I assume it's for non-fused model right? The change does not seem to touch contrib ops

Yes. In BERT-L, we only fused Layernorm, GELU, Transpose+Matmul.

weixingzhang commented:
> In general, I have concerns of regression in this change not caught by existing tests. I had added comments on some ops that fixed sized TArray may overflow.
Agree, I just updated PR to keep CudaAsyncBuffer for non_max_suppression, cudnn_rnn_base, concat, split. Thanks!

weixingzhang commented:
> In general, I have concerns of regression in this change not caught by existing tests. I had added comments on some ops that fixed sized TArray may overflow.

Agree, I just updated PR to keep CudaAsyncBuffer for non_max_suppression, cudnn_rnn_base, concat, split. Thanks!
